src/components/JobRunner.js

```diff
@@ -1,160 +1,364 @@
-import ReactJson from 'react-json-view'
-
-const RunResponse = ({data, ...props}) => {
-    return (
-        <div >
-            <ReactJson src={data} displayObjectSize={false} displayDataTypes={false}/>
-        </div>
-    );
-};
+import React, { Component } from "react";
+
+import RunResponse from "./RunResponse";
+import TestResponse from "./TestResponse";
 
-const TestResponse = ({data, ...props}) => {
-    return (
-        <div >
-            <ReactJson src={data} displayObjectSize={false} displayDataTypes={false}/>
-        </div>
-    );
+import LimitForm from "./Forms/LimitForm";
+import PhenotypeForm from "./Forms/PhenotypeForm";
+import DocumentSetForm from "./Forms/DocumentSetForm";
+import TermsetForm from "./Forms/TermsetForm";
+import CohortForm from "./Forms/CohortForm";
+import DefineFeatureForm from "./Forms/DefineFeatureForm";
+import LogicalContextForm from "./Forms/LogicalContextForm";
+import DefineResultForm from "./Forms/DefineResultForm";
+
+const initialState = {
+    editing: false,
+    editText: "Edit",
+    phenotypeModal: "is-active",
+    termSets: [],
+    documentSets: [],
+    cohorts: [],
+    features: [],
+    response_view: null,
+    limitModal: null
 };
 
 class JobRunner extends Component {
+    constructor(props) {
+        super(props);
+
+        this.state = initialState;
+    }
 
+    response_ERROR = () => {
+        return (
+            <TestResponse
+                valid={false}
+                data={{
+                    ERROR: "There was an error with the nlpql."
+                }}
+                toggle={this.toggleResponse}
+                toggleLimitModal={this.toggleLimitModal}
+            />
+        );
+    };
 
+    getNamesArray = arr => {
+        return arr.map(value => {
+            return value.name;
+        });
+    };
 
-    constructor(props) {
-        super(props);
-        this.base_url = props.url;
-
-        this.toggle = this.toggle.bind(this);
-        this.getNLPQLSample = this.getNLPQLSample.bind(this);
-        this.updateNLPQL = this.updateNLPQL.bind(this);
-        this.handleButtonAction = this.handleButtonAction.bind(this);
-        this.clear = this.clear.bind(this);
-
-        this.state = {
-            dropdownOpen: false,
-            nlpql: '',
-            test_response: {},
-            run_response: {}
-        };
+    getFeaturesArray = arr => {
+        return arr.map(value => {
+            return {
+                name: value.name,
+                algorithm: value.funct
+            };
+        });
+    };
+
+    setArraysFromJSON = () => {
+        if (!this.props.runner.nlpql_JSON) return;
+
+        let termSets = [];
+        let documentSets = [];
+        let cohorts = [];
+        let features = [];
+
+        this.props
+            .postToClarityAPI("nlpql_tester", this.props.runner.nlpql)
+            .then(() => {
+                if (this.props.runner.nlpql_JSON.term_sets) {
+                    termSets = this.getNamesArray(
```

Comment:
because of the further computation that is necessary on these values, it cannot be done within the same state update. Since we are not using these arrays often, I have opted to keep them as a state and pass them through props rather than implement an extra action to update them.